### ISSUE TITLE
Multiple targets per service (load balancing)

### ISSUE THREAD

**Issue by kevinmcconnell** [basecamp/kamal-proxy#15]
We should add the ability to register multiple targets for a single service, and distribute requests amongst them round-robin. This will give us the ability to load balance the traffic for a service across multiple hosts. With this setup, Kamal can provide a complete solution for running an app over multiple hosts, by running a proxy instance on a different host.

Adding this also means:

- The semantics of deploying needs to be extended to handle the case where only some of the new targets become healthy. Probably the behaviour here should be that we health-check every host in the list, and if any of them fail to become healthy we return non-zero without having applied any of the changes. But if all become healthy, we apply the new set as an atomic change. This includes the case there there are overlaps between the old and new list.

  For example, if you're deployed to t1, t2, t3 and you attempt a new deployment for t1, t2, t5, t6 but t6 fails, we keep the original set of t1, t2, t3.
- Healthchecks should be ongoing, rather than stopping after success as they do now. We'll want to notice an unhealthy target after the fact and take it out of the routing until it becomes healthy again.

**Comment by Jainam-17-18** [comment]
Can we have internal load balancer in Kamal-proxy which can handle request by forwarding request to multiple application containers running on **Same Server** or **Single Host**. And can we have Kamal deploy config like ```replicas:  5 ```(container count), so total 5 container will be created during deploy and Kamal-proxy will handle it.

![Image](https://github.com/user-attachments/assets/765b556f-be89-4ced-a221-3991cc37c269)

**Comment by mhenrixon** [comment]
@kevinmcconnell @djmb have you guys thought more about this? I just ran into a project where I could use load balancing to avoid resorting to traefik.

**Comment by kevinmcconnell** [comment]
@mhenrixon yes, this is still definitely planned. I recently started work on it, and have a branch that's close to ready. Hope to have something pushed up to share soon.

At the proxy level, it adds the ability to specify multiple `target` params, and implements the behaviour described above. So if you do something like `deploy myservice --target=app01,app02,app03` then it will do round-robin load balancing between those three servers (moving them in and out of rotation according to their healthchecks).

We're still discussing the details of how this will look in Kamal's `deploy.yml`, but again hopefully will have more to share soon on that!

**Comment by mhenrixon** [comment]
Oh shoot @kevinmcconnell I kind of got into it and didn't see your message until now. 

If you already have something you can close my PR. I have like 3 weeks until I need it but I have a fallback ready with to Traefik (or even a Hetzner load balancer would work I guess).

Let me know if I can help in any way or form.

**Comment by kevinmcconnell** [comment]
Thanks for your help with this, @mhenrixon! I'm going to close your PR in favor of [the one I've had brewing](https://github.com/basecamp/kamal-proxy/pull/124), which I just opened. But I do appreciate you taking a shot at this! Looks like there are quite a few places where we took a similar approach.

I'm planning to merge my PR fairly soon, but if you have a chance to try that branch on your project I'd love to hear how you get on with it, and any feedback you have.

**Comment by mhenrixon** [comment]
> I'm planning to merge my PR fairly soon, but if you have a chance to try that branch on your project I'd love to hear how you get on with it, and any feedback you have.

I'd love to try it out. 

Can I take some shortcuts to deploy my app with the binaries built from your branch? If you could help me with how to approach this, I'll be sure to get it done today @kevinmcconnell. 

This is simply too good to not prioritize!

**Comment by kevinmcconnell** [comment]
@mhenrixon as it's not integrated into Kamal yet, you won't be able to get it through a normal `kamal deploy`. But I've pushed this version to Docker Hub, [tagged as `lb`](https://hub.docker.com/layers/basecamp/kamal-proxy/lb/images/sha256-9f19c2adbeee8528614ed213146728bc80904e401d3cdbd826ed89db6553ea4a), so you could try running it directly with Docker. Something like:

```
docker run --name load-balancer basecamp/kamal-proxy:lb
```

And then you can deploy your app's hosts into it by running `kamal-proxy` commands against that container, like:

```
docker exec load-balancer kamal-proxy deploy myapp --target=host01,host02,host03
```

But of course you can also build it locally from the branch too. Just run `make` to create a plain binary, or `make docker` to build your own image. Any of these ways should work.

**Comment by mhenrixon** [comment]
Ok, if I use the new proxy with the lb tag on each server it looks like I get connection refused:

```json
{
    "time": "2025-04-10T10:09:05.009555624Z",
    "level": "INFO",
    "msg": "Healthcheck failed",
    "url": "http://IP_1/up",
    "error": "Get \"http://IP_1/up\": dial tcp 65.108.243.92:80: connect: connection refused"
  },
  {
    "time": "2025-04-10T10:09:05.010584576Z",
    "level": "INFO",
    "msg": "Healthcheck failed",
    "url": "http://IP_2/up",
    "error": "Get \"http://IP_2/up\": dial tcp 157.180.40.129:80: connect: connection refused"
  }
```

**Comment by mhenrixon** [comment]
Ok that worked!

```json
{
  "time": "2025-04-10T14:18:07.684601197Z",
  "level": "INFO",
  "msg": "Deployed",
  "service": "wm3",
  "hosts": [
    // ... list of domains
  ],
  "paths": [
    "/"
  ],
  "targets": [
    // ... list of ip
  ],
  "tls": true,
  "strip": true
}
```

So the trick was to use port 80 since the Rails ports are not responding, which is correct if the regular kamal-proxy is supposed to respond- which it should, to improve serving files with thruster, correct?

Anyway. my requests are not hitting the app servers. They seem to hang forever after it has booted. 

It is almost like it is stuck just checking the health checks and forgetting to respond to requests. I see continuous health checks on the web servers, but nothing is logged after this final log entry that I posted above.

**Comment by kevinmcconnell** [comment]
Ah, maybe the load balancer isn't accessible externally? You will need to publish its ports when you run it (I forgot to mention that earlier). If you haven't already, could you try starting it like this:

```
docker run --name load-balancer -p 443:443 -p 80:80 basecamp/kamal-proxy:lb
```

**Comment by mhenrixon** [comment]
Doh 🤦‍♂️ of course, I have to publish the ports! 

```json
{
  "time": "2025-04-10T16:39:39.265976849Z",
  "level": "INFO",
  "msg": "Request",
  "host": "LB_IP:80",
  "port": 80,
  "path": "/",
  "request_id": "fd4d2049-4d2b-4ffd-975e-b351450edabb",
  "status": 404,
  "service": "",
  "target": "",
  "duration": 183967,
  "method": "GET",
  "req_content_length": 0,
  "req_content_type": "application/x-www-form-urlencoded; charset=UTF-8",
  "resp_content_length": 4492,
  "resp_content_type": "text/html; charset=utf-8",
  "client_addr": "UPSTREAM_IP",
  "client_port": "55622",
  "remote_addr": "UPSTREAM_IP",
  "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.85 Safari/537.36 Edg/90.0.818.46",
  "proto": "HTTP/1.1",
  "scheme": "http",
  "query": ""
}
```

Can I do anything to help finish this on the kamal side? I sincerely need this

**Comment by kevinmcconnell** [comment]
@mhenrixon so it's working for you properly now? (Other than having to do the manual setup, of course). I noticed your log line there is a 404 -- was that  a problem you were hitting, or just a request that landed on the load balancer while it didn't have anything deployed into it?

> Can I do anything to help finish this on the kamal side?

I'll catch up with @djmb (main Kamal maintainer) about his plans for that side, as I know he has some ideas about the best way to integrate this. Then we can figure out the next steps. We're keen to make adding a load balancer to a Kamal deployment as simple as possible, so it'll require a bit of thought about how best to incorporate all the details into the deploy.yml.

In the meantime any feedback from real-world use of the proxy part would be helpful in confirming that we've got all the bits we need there. So if you are using this (even if with a bit of manual setup) it would be great to hear how you get on. And in any case, we'll try to get things moving on the Kamal side soon!

In the meantime I'm going to merge my branch to main. It might still evolve a bit from here, but I think the API should be pretty stable at this point.

**Comment by mhenrixon** [comment]
> We're keen to make adding a load balancer to a Kamal deployment as simple as possible, so it'll require a bit of thought about how best to incorporate all the details into the deploy.yml.

Actually, I am working on something for this. I see what you mean; there are some nuances with the setup. I wanted to make it completely seamless so that if more than one host for the web container is present, the first one gets the load balancer, and then you can specify one. 

https://github.com/basecamp/kamal/pull/1490 -- This PR works for me, and I am currently deploying the `:lb` tag with it. 

I am sure you guys can clean it up massively, but I am pleased with being able to use the load balancer already.

**Comment by kevinmcconnell** [comment]
@mhenrixon that's great, thanks for getting this started! I appreciate your help with this, and I'm glad the load balancing feature is proving useful to you!

**Comment by mhenrixon** [comment]
> Is this feature available in kamal ?

Nope, only through my PR but it is working nicely in a few projects for me.
